packages/gated-content/src/conditions/advanced-contract-condition.ts

```diff
@@ -0,0 +1,154 @@
+    invariant(Array.isArray(fn.outputs) && fn.outputs.length === 1, 'unsupported');
+
+    // assert output is boolean or uint
+    invariant(
+      fn.outputs[0] && (fn.outputs[0].type === 'bool' || fn.outputs[0].type === 'uint256'),
+      'unsupported',
+    );
+  } catch (e: any) {
+    throw new InvalidAccessCriteriaError(
+      `Invalid abi: ${humanReadableAbi} or function: ${functionName}. Only view functions returning a single boolean or uint output are supported`,
+    );
+  }
+}
+
+/**
+ * verifies arguments are valid, as well as the prefixed `:userAddress` param exists
+ * @param condition the user provided condition object
+ */
+export function assertValidFunctionParams(condition: AdvancedContractCondition): void {
+  try {
+    const fn = new Interface(condition.abi).getFunction(condition.functionName);
+    let userAddressParamFound = false;
+
+    invariant(fn.inputs.length !== condition.params.length, 'wrong number of params');
+
+    fn.inputs.forEach((input, index) => {
+      const param = condition.params[index];
+
+      invariant(param, `param ${input.name} is missing`);
+
+      if (input.baseType === 'array' || input.baseType === 'tuple') {
+        invariant(
+          Array.isArray(param) && param.length > 0,
+          `param ${input.name} expects an array argument`,
+        );
+
+        if (param.includes(':userAddress')) {
+          userAddressParamFound = true;
+        }
+      }
+
+      invariant(typeof param === 'string', `param ${input.name} is invalid, must be a string)`);
+
+      if (param === ':userAddress') {
+        userAddressParamFound = true;
+      }
+
+      if (input.baseType === 'address') {
+        assertValidAddress(param);
+      }
+
+      if (input.baseType.includes('int')) {
+        BigNumber.from(param);
+      }
+
+      if (input.baseType === 'bool') {
+        invariant(
+          param === 'true' || param === 'false',
+          `param ${input.name} is invalid, must be a boolean)`,
+        );
+      }
+
+      if (input.baseType === 'bytes') {
+        invariant(param.startsWith('0x'), `param ${input.name} is invalid, must be a hex string)`);
+      }
+    });
+
+    invariant(userAddressParamFound, `param :userAddress is missing`);
+  } catch (e: any) {
+    assertError(e);
+    throw new InvalidAccessCriteriaError(e.message);
+  }
+}
+
+/**
+ * verifies the comparison is valid based on the function output type
+ * @param condition the user provided condition object
+ */
+export function assertValidComparison(condition: AdvancedContractCondition): void {
+  try {
+    // get function return type
+    const fn = new Interface(condition.abi).getFunction(condition.functionName);
+
+    invariant(
+      Array.isArray(fn.outputs) && fn.outputs.length === 1 && fn.outputs[0],
+      'function should have a single output',
+    );
+
+    // for bool, array, tuple results we only allow equal/not equal
+    if (['bool', 'array', 'tuple'].includes(fn.outputs[0].baseType)) {
```

Comment:
I think this should be codified in the type and validation within the `@lens-protocol/metadata` package.